### PR TITLE
Adding function Subsets to the mathics

### PR DESCRIPTION
**Pull request by larionvn**
Adding Subsets[]

### CODE REVIEW

File: mathics/builtin/combinatorial.py [hunk 1]
```diff
@@ -270,3 +270,172 @@ class RogersTanimotoDissimilarity(_BooleanDissimilarity):
...
+    def apply_1(self, list, n, evaluation):
+        'Subsets[list_?ListQ, n_]'
+        
+        tlist = [x for x in list.leaves]
+        expr = Expression('Subsets', list, n)
+        
+        if not n.get_head_name() == 'System`Integer' or n.to_python() < 0 :
+            return evaluation.message('Subsets', 'nninfseq', expr)
+        
+        tlen = int(n.to_python())
```
Comment (sn6uv) [inline mathics/builtin/combinatorial.py]: we have a utility method for this, `n.get_int_value()` which returns an int or None.

File: mathics/builtin/combinatorial.py [hunk 2]
```diff
@@ -270,3 +270,172 @@ class RogersTanimotoDissimilarity(_BooleanDissimilarity):
...
+        
+        nested_list = [Expression('List', *c) for i in range(0, tlen + 1) for c in combinations(tlist, i)]
```
Comment (sn6uv) [inline mathics/builtin/combinatorial.py]: `range(0, tlen +1)` is `range(tlen+1)`

File: mathics/builtin/combinatorial.py [hunk 3]
```diff
@@ -270,3 +270,172 @@ class RogersTanimotoDissimilarity(_BooleanDissimilarity):
...
+        
+        if n_len > 3:
+            return evaluation.message('Subsets', 'nninfseq', expr)
+        
+        if n_len == 0:
+            return evaluation.message('Subsets', 'nninfseq', expr)
+        
+        if n_len == 1:
+            elem1 = n_python[0]
+            if not n.leaves[0].get_head_name() == "System`Integer" or elem1 < 0 :
```
Comment (sn6uv) [inline mathics/builtin/combinatorial.py]: Not all values can be converted to_python so elem1 may be None here. Better to use `get_int_value()` here too.

File: mathics/builtin/combinatorial.py [hunk 4]
```diff
@@ -270,3 +270,172 @@ class RogersTanimotoDissimilarity(_BooleanDissimilarity):
...
+                return evaluation.message('Subsets', 'nninfseq', expr)
+            min_n = elem1
+            max_n = min_n + 1
+            step_n = 1
+        
+        if n_len == 2:
+            elem1 = n_python[0]
+            elem2 = n_python[1]
+            if not n.leaves[0].get_head_name() == "System`Integer" or not n.leaves[1].get_head_name() == "System`Integer" or elem1 < 0 or elem2 < 0 :
```
Comment (sn6uv) [inline mathics/builtin/combinatorial.py]: `get_int_value()`

File: mathics/builtin/combinatorial.py [hunk 5]
```diff
@@ -270,3 +270,172 @@ class RogersTanimotoDissimilarity(_BooleanDissimilarity):
...
+                return evaluation.message('Subsets', 'nninfseq', expr)
+            min_n = elem1
+            max_n = elem2 + 1
+            step_n = 1
+            
+        if n_len == 3:
+            if not n.leaves[0].get_head_name() == "System`Integer" or not n.leaves[1].get_head_name() == "System`Integer" or not n.leaves[2].get_head_name() == "System`Integer" :
```
Comment (sn6uv) [inline mathics/builtin/combinatorial.py]: `get_int_value()`

File: mathics/builtin/combinatorial.py [hunk 6]
```diff
@@ -270,3 +270,172 @@ class RogersTanimotoDissimilarity(_BooleanDissimilarity):
...
+    
+        return self.apply_1(list, Integer(len(list.to_python())), evaluation)
+    
+    def apply_1(self, list, n, evaluation):
+        'Subsets[list_?ListQ, n_]'
+        
+        tlist = [x for x in list.leaves]
+        expr = Expression('Subsets', list, n)
+        
+        if not n.get_head_name() == 'System`Integer' or n.to_python() < 0 :
```
Comment (sn6uv) [inline mathics/builtin/combinatorial.py]: `get_int_value()`, see below.

**Comment by larionvn** [comment]
Thanks for your feedback @sn6uv . I've just pushed the latest commit for adding Subsets[] with some changes  as your recommendations. Please take a look and give me some feedback if any. Thanks in advance!

File: mathics/builtin/combinatorial.py [hunk 7]
```diff
@@ -370,32 +373,27 @@ def apply(self, list, evaluation):
...
         'Subsets[list_?ListQ, n_]'
         
-        tlist = [x for x in list.leaves]
         expr = Expression('Subsets', list, n)
         
-        if not n.get_head_name() == 'System`Integer' or n.to_python() < 0 :
+        n_value = n.get_int_value() 
+        if n_value == 0:
+            return Expression('List', Expression('List'))
+        if n_value is None or n_value < 0:
```
Comment (sn6uv) [inline mathics/builtin/combinatorial.py]: I think the order of these if statements should be swapped. Can you add a test, e.g. `Subsets[{1, 2}, x]`

File: mathics/builtin/combinatorial.py [hunk 8]
```diff
@@ -370,32 +373,27 @@ def apply(self, list, evaluation):
...
             return evaluation.message('Subsets', 'nninfseq', expr)
         
-        tlen = int(n.to_python())
+        nested_list = [Expression('List', *c) for i in range(n_value + 1) for c in combinations([x for x in list.leaves], i)]
```
Comment (sn6uv) [inline mathics/builtin/combinatorial.py]: `... for c in combinations(list.leaves, ....`?

File: mathics/builtin/combinatorial.py [hunk 9]
```diff
@@ -370,32 +373,27 @@ def apply(self, list, evaluation):
...
-        tlist = [x for x in list.leaves]
         expr = Expression('Subsets', list, n)
         
         if n.get_name() == 'System`All' or n.has_form('DirectedInfinity', 1):
-            tlen = len(tlist)
             return self.apply(list, evaluation)
         
-        n_python = n.to_python()
-        n_len =  len(n_python)
+        n_len = n.leaves.__len__()
```
Comment (sn6uv) [inline mathics/builtin/combinatorial.py]: `len(n.leaves)`?

File: mathics/builtin/combinatorial.py [hunk 10]
```diff
@@ -404,38 +402,39 @@ def apply_2(self, list, n, evaluation):
             return evaluation.message('Subsets', 'nninfseq', expr)
         
         if n_len == 1:
-            elem1 = n_python[0]
-            if not n.leaves[0].get_head_name() == "System`Integer" or elem1 < 0 :
+            elem1 = n.leaves[0].get_int_value()
+            if not elem1 or elem1 < 0 :
```
Comment (sn6uv) [inline mathics/builtin/combinatorial.py]: `not elem1` will also be true if `elem1` is `0`, use `elem1 is not None`

File: mathics/builtin/combinatorial.py [hunk 11]
```diff
@@ -404,38 +402,39 @@ def apply_2(self, list, n, evaluation):
...
             elif step_n < 0:
-                min_n = n_python[0]
-                max_n = n_python[1] - 1
+                min_n = elem1
+                max_n = elem2 - 1
             else:
                 return evaluation.message('Subsets', 'nninfseq', expr)
             
-        nested_list = [Expression('List', *c) for i in range(min_n, max_n, step_n) for c in combinations(tlist, i)]
+        nested_list = [Expression('List', *c) for i in range(min_n, max_n, step_n) for c in combinations([x for x in list.leaves], i)]
```
Comment (sn6uv) [inline mathics/builtin/combinatorial.py]: `combinations(list.leaves...` here too

File: mathics/builtin/combinatorial.py [hunk 12]
```diff
@@ -270,3 +270,171 @@ class RogersTanimotoDissimilarity(_BooleanDissimilarity):
...
+        
+        if n.get_name() == 'System`All' or n.has_form('DirectedInfinity', 1):
+            return self.apply(list, evaluation)
+        
+        n_len = n.leaves.__len__()
+        
+        if n_len > 3:
+            return evaluation.message('Subsets', 'nninfseq', expr)
+        
+        if n_len == 0:
```
Comment (Piruzzolo) [inline mathics/builtin/combinatorial.py]: better an ELIF?

File: mathics/builtin/combinatorial.py [hunk 13]
```diff
@@ -270,3 +270,198 @@ class RogersTanimotoDissimilarity(_BooleanDissimilarity):
...
+                    return evaluation.message('Subsets', 'nninfseq', expr)
+                min_n = elem1
+                max_n = elem2 + 1
+                step_n = 1
+                
+            elif n_len == 3:
+                elem1 = n.leaves[0].get_int_value()
+                elem2 = n.leaves[1].get_int_value()
+                elem3 = n.leaves[2].get_int_value()
+                if elem1 is None or elem2 is None or elem3 is None :
```
Comment (sn6uv) [inline mathics/builtin/combinatorial.py]: Should we also be checking that `elem1` and `elem2` aren't negative?

File: mathics/builtin/combinatorial.py [hunk 14]
```diff
@@ -270,3 +270,198 @@ class RogersTanimotoDissimilarity(_BooleanDissimilarity):
...
+            elif n_len == 2:
+                elem1 = n.leaves[0].get_int_value()
+                elem2 = n.leaves[1].get_int_value()
+                if elem1 is None or elem2 is None or elem1 < 0 or elem2 < 0 :
+                    return evaluation.message('Subsets', 'nninfseq', expr)
+                min_n = elem1
+                max_n = elem2 + 1
+                step_n = 1
+                
+            elif n_len == 3:
```
Comment (sn6uv) [inline mathics/builtin/combinatorial.py]: docs for this case missing?

**Comment by larionvn** [comment]
Thanks for your feedback @sn6uv . I've just pushed the latest source code with the following changes:
 - Update code so that Subsets can find any subsets from any head
 - Adding some test cases and docs as your recommendations.

File: mathics/builtin/combinatorial.py [hunk 15]
```diff
@@ -440,9 +470,9 @@ def apply_2(self, list, n, evaluation):
                 
             elif n_len == 3:
                 elem1 = n.leaves[0].get_int_value()
-                elem2 = n.leaves[1].get_int_value()
+                elem2 = n.leaves[1].get_int_value() if not n.leaves[1].has_form('DirectedInfinity', 1) else len(list.leaves) + 1
                 elem3 = n.leaves[2].get_int_value()
-                if elem1 is None or elem2 is None or elem3 is None :
+                if elem1 is None or elem2 is None or elem3 is None or elem1 < 0 or elem2 < 0:
```
Comment (sn6uv) [inline mathics/builtin/combinatorial.py]: Does this handle all the cases, e.g. `elem3 <= 0`, `elem2 > elem1`?

Comment (larionvn) [inline mathics/builtin/combinatorial.py]: Yes, it does. In the above case, the range(min_n, max_n, step_n) returns an empty list [] so the final result will be an empty list {}.

**Comment by rocky** [comment]
I am not seeing anything left unresolved. What am I missing?

**Pull request merged**